test/e2e/service.go

```diff
@@ -437,54 +539,95 @@ func validateUniqueOrFail(s []string) {
 		for _, port := range ss.Ports {
-			if port.Port == expectedPort {
+			if _, ok := expectedPorts[port.Port]; ok {
 				for _, addr := range ss.Addresses {
-					ips.Insert(addr.IP)
+					Logf("Found IP %v and port %v", addr.IP, port.Port)
+					if _, ok = m[addr.IP]; !ok {
+						m[addr.IP] = make([]int, 0)
+					}
+					m[addr.IP] = append(m[addr.IP], port.Port)
 				}
 			}
 		}
 	}
-	return ips
+	return m
 }
 
-func validateIPsOrFail(c *client.Client, ns string, expectedPods []string, ips util.StringSet) {
-	for _, name := range expectedPods {
+func translatePodNameToIpOfFail(c *client.Client, ns string, expectedEndpoints map[string][]int) map[string][]int {
+	portsByIp := make(map[string][]int)
+
+	for name, portList := range expectedEndpoints {
 		pod, err := c.Pods(ns).Get(name)
 		if err != nil {
 			Failf("failed to get pod %s, that's pretty weird. validation failed: %s", name, err)
 		}
-		if !ips.Has(pod.Status.PodIP) {
-			Failf("ip validation failed, expected: %v, saw: %v", ips, pod.Status.PodIP)
-		}
+		portsByIp[pod.Status.PodIP] = portList
 		By(fmt.Sprintf(""))
 	}
-	By(fmt.Sprintf("successfully validated IPs %v against expected endpoints %v on namespace %s", ips, expectedPods, ns))
+	By(fmt.Sprintf("successfully translated pod names to ips: %v -> %v on namespace %s", expectedEndpoints, portsByIp, ns))
+	return portsByIp
 }
 
-func validateEndpointsOrFail(c *client.Client, ns, serviceName string, expectedPort int, expectedPods []string) {
+func validatePortsOrFail(endpoints map[string][]int, expectedEndpoints map[string][]int) {
+	if len(endpoints) != len(expectedEndpoints) {
+		Failf("invalid number of endpoints (should not happen because we check this condition before) got %v, expected %v", endpoints, expectedEndpoints)
+	}
+	for ip := range expectedEndpoints {
+		if _, ok := endpoints[ip]; !ok {
+			Failf("endpoint %v not found", ip)
+		}
+		if len(endpoints[ip]) != len(expectedEndpoints[ip]) {
+			Failf("invalid list of ports for ip %v. Got %v, expected %v", ip, endpoints[ip], expectedEndpoints[ip])
+		}
+		sort.Ints(endpoints[ip])
+		sort.Ints(expectedEndpoints[ip])
+		for index := range endpoints[ip] {
+			if endpoints[ip][index] != expectedEndpoints[ip][index] {
+				Failf("invalid list of ports for ip %v. Got %v, expected %v", ip, endpoints[ip], expectedEndpoints[ip])
+			}
+		}
+	}
+}
+
+func validateEndpointsOrFail(c *client.Client, ns, serviceName string, expectedEndpoints map[string][]int) {
+	expectedPorts := make(map[int]bool)
+	for _, ports := range expectedEndpoints {
+		for _, port := range ports {
+			expectedPorts[port] = true
+		}
+	}
+
 	for {
 		endpoints, err := c.Endpoints(ns).Get(serviceName)
 		if err == nil {
-			ips := flattenSubsets(endpoints.Subsets, expectedPort)
-			if len(ips) == len(expectedPods) {
-				validateIPsOrFail(c, ns, expectedPods, ips)
+			By(fmt.Sprintf("Found endpoints %v", endpoints))
+
+			// I don't really understand why we are ignoring ports that are not expected to be present.
```

Comment:
As discussed, I removed ignoring unknown ports. Now the test will fail if we receive an unexpected port (endpoint) in the service, which is better.